docs/specs/lcb-traits.md

```diff
@@ -0,0 +1,302 @@
+   the whole of LiveCode.  This provides an extremely high barrier to entry if
+   all that's wanted is to add a type specialization to a syntax that is
+   already standard.
+
+3. The only "interesting" part of the specialization is the handler which
+   performs the formatting.  The rest of the definition is redundant.
+
+### Generic handlers are insufficient
+
+It has previously been suggested that Builder should have generic handlers.
+For example, the standard library could contain a generic handler declaration:
+
+    module S
+
+    public generic handler FormatAsString(in pOperand as generic, \
+                                          out rFormat as string)
+
+Then, another module could define a formattable type:
+
+    module M
+    use S
+    public type T -- Some full declaration
+
+    public handler FormatAsString(in pOperand as T, out rFormat as string)
+        -- Perform formatting on T
+    end handler
+
+In this case, the compiler would identify that `M.FormatAsString<T>()` has a
+name that matches a generic handler declaration, and that it conforms (all
+generic parameters are specialised and all non-generic parameters are of the
+correct types).  Subsequent calls to `FormatAsString()` from *anywhere in the
+program* would then be able to access `M.FormatAsString<T>()` if the first
+argument of `FormatAsString()` conforms to `T`.
+
+The main drawback here is that Builder is optionally-typed, but it would be
+good to ensure that it's possible to write code where all types are specified
+*and can be verified* at compile time.  This doesn't satisfy the requirement.
+
+For example, it would be convenient to write a `Log()` handler that can log
+any type that can also be formatted.  However, with the generic handlers scheme
+described above, `Log()` would have to take arguments as `any` and defer
+type-checking to runtime.
+
+## Traits
+
+Traits are collections of methods that can be used to extend the functionality
+of a type.
+
+Traits have no storage associated with them.  Storage is provided by
+implementing the trait for a specific storage type.
+
+### Defining traits
+
+    public trait type CanFormatAsString
+        method type FormatAsString(in me) as string
+    end trait
+
+A trait definition is a block containing a list zero or more `method`
+definitions or `method type` declarations.  `method type` declarations take the
+same form as existing `handler type` declarations.  `method` definitions take
+the same form as existing `handler` declarations.
+
+    public trait type CanFormatAsString
+        method FormatAsString(in me) as string
+            return "<unknown>"
+        end method
+    end trait
+
+`method` definitions allow a trait's definition to include default
+implementations of the trait's methods.  Default implementations can call other
+methods of the trait or perform any operations valid for the generic `Self`
+type.
+
+Two additional symbols are defined within a trait definition:
```

Comment:
That wasn't quite what I meant :)

I was suggesting that trait methods don't necessarily need an explicit 'me' - we could use the same formalism as object methods in other languages where me/self/this is implicit.

If we did add the equivalent of 'static' methods it would probably make more sense to have it as an explicit type of method rather than implicit from the argument list.